exchanges/coinbasepro/coinbasepro_wrapper.go

```diff
@@ -871,26 +905,502 @@ func (c *CoinbasePro) ValidateAPICredentials(ctx context.Context, assetType asse
 		return time.Time{}, err
 	}
-	return st.ISO, nil
+	return st.Data.ISO, nil
 }
 
 // GetLatestFundingRates returns the latest funding rates data
-func (c *CoinbasePro) GetLatestFundingRates(context.Context, *fundingrate.LatestRateRequest) ([]fundingrate.LatestRateResponse, error) {
-	return nil, common.ErrFunctionNotSupported
+func (c *CoinbasePro) GetLatestFundingRates(ctx context.Context, r *fundingrate.LatestRateRequest) ([]fundingrate.LatestRateResponse, error) {
+	if r == nil {
+		return nil, common.ErrNilPointer
+	}
+	if !c.SupportsAsset(r.Asset) {
+		return nil, fmt.Errorf("%w %v", asset.ErrNotSupported, r.Asset)
+	}
+	verified, err := c.verificationCheck(ctx)
+	if err != nil {
+		return nil, err
+	}
+	products, err := c.fetchFutures(ctx, verified)
+	if err != nil {
+		return nil, err
+	}
+	funding := make([]fundingrate.LatestRateResponse, len(products.Products))
+	for i := range products.Products {
+		pair, err := currency.NewPairFromString(products.Products[i].ID)
+		if err != nil {
+			return nil, err
+		}
+		funRate := fundingrate.Rate{Time: products.Products[i].FutureProductDetails.PerpetualDetails.FundingTime,
+			Rate: decimal.NewFromFloat(products.Products[i].FutureProductDetails.PerpetualDetails.FundingRate.Float64()),
+		}
+		funding[i] = fundingrate.LatestRateResponse{
+			Exchange:    c.Name,
+			Asset:       r.Asset,
+			Pair:        pair,
+			LatestRate:  funRate,
+			TimeChecked: time.Now(),
+		}
+	}
+	return funding, nil
 }
 
 // GetFuturesContractDetails returns all contracts from the exchange by asset type
-func (c *CoinbasePro) GetFuturesContractDetails(context.Context, asset.Item) ([]futures.Contract, error) {
-	return nil, common.ErrFunctionNotSupported
+func (c *CoinbasePro) GetFuturesContractDetails(ctx context.Context, item asset.Item) ([]futures.Contract, error) {
+	if !item.IsFutures() {
+		return nil, futures.ErrNotFuturesAsset
+	}
+	if !c.SupportsAsset(item) {
+		return nil, fmt.Errorf("%w %v", asset.ErrNotSupported, item)
+	}
+	verified, err := c.verificationCheck(ctx)
+	if err != nil {
+		return nil, err
+	}
+	products, err := c.fetchFutures(ctx, verified)
+	if err != nil {
+		return nil, err
+	}
+	contracts := make([]futures.Contract, len(products.Products))
+	for i := range products.Products {
+		pair, err := currency.NewPairFromString(products.Products[i].ID)
+		if err != nil {
+			return nil, err
+		}
+		funRate := fundingrate.Rate{Time: products.Products[i].FutureProductDetails.PerpetualDetails.FundingTime,
+			Rate: decimal.NewFromFloat(products.Products[i].FutureProductDetails.PerpetualDetails.FundingRate.Float64()),
+		}
+		contracts[i] = futures.Contract{
+			Exchange:             c.Name,
+			Name:                 pair,
+			Asset:                item,
+			EndDate:              products.Products[i].FutureProductDetails.ContractExpiry,
+			IsActive:             !products.Products[i].IsDisabled,
+			Status:               products.Products[i].Status,
+			Type:                 futures.LongDated,
```

Comment:
You function `fetchFutures` returns both delivery futures and perpetual futures. You will need to distinguish between that for these types.
A perpetual future does not have a long dated expiry